commons-math/src/main/java/org/apache/commons/math4/optim/nonlinear/scalar/noderiv/BOBYQAOptimizer.java

```diff
@@ -0,0 +1,2475 @@
+                    final double deltaTwo = -initialTrustRegionRadius;
+                    lowerDifference.setEntry(j, FastMath.min(deltaOne, deltaTwo));
+                    upperDifference.setEntry(j, initialTrustRegionRadius);
+                }
+            }
+        }
+
+        // Make the call of BOBYQB.
+
+        return bobyqb(lowerBound, upperBound);
+    } // bobyqa
+
+    // ----------------------------------------------------------------------------------------
+
+    /**
+     *     The arguments N, NPT, X, XL, XU, RHOBEG, RHOEND, IPRINT and MAXFUN
+     *       are identical to the corresponding arguments in SUBROUTINE BOBYQA.
+     *     XBASE holds a shift of origin that should reduce the contributions
+     *       from rounding errors to values of the model and Lagrange functions.
+     *     XPT is a two-dimensional array that holds the coordinates of the
+     *       interpolation points relative to XBASE.
+     *     FVAL holds the values of F at the interpolation points.
+     *     XOPT is set to the displacement from XBASE of the trust region centre.
+     *     GOPT holds the gradient of the quadratic model at XBASE+XOPT.
+     *     HQ holds the explicit second derivatives of the quadratic model.
+     *     PQ contains the parameters of the implicit second derivatives of the
+     *       quadratic model.
+     *     BMAT holds the last N columns of H.
+     *     ZMAT holds the factorization of the leading NPT by NPT submatrix of H,
+     *       this factorization being ZMAT times ZMAT^T, which provides both the
+     *       correct rank and positive semi-definiteness.
+     *     NDIM is the first dimension of BMAT and has the value NPT+N.
+     *     SL and SU hold the differences XL-XBASE and XU-XBASE, respectively.
+     *       All the components of every XOPT are going to satisfy the bounds
+     *       SL(I) .LEQ. XOPT(I) .LEQ. SU(I), with appropriate equalities when
+     *       XOPT is on a constraint boundary.
+     *     XNEW is chosen by SUBROUTINE TRSBOX or ALTMOV. Usually XBASE+XNEW is the
+     *       vector of variables for the next call of CALFUN. XNEW also satisfies
+     *       the SL and SU constraints in the way that has just been mentioned.
+     *     XALT is an alternative to XNEW, chosen by ALTMOV, that may replace XNEW
+     *       in order to increase the denominator in the updating of UPDATE.
+     *     D is reserved for a trial step from XOPT, which is usually XNEW-XOPT.
+     *     VLAG contains the values of the Lagrange functions at a new point X.
+     *       They are part of a product that requires VLAG to be of length NDIM.
+     *     W is a one-dimensional array that is used for working space. Its length
+     *       must be at least 3*NDIM = 3*(NPT+N).
+     *
+     * @param lowerBound Lower bounds.
+     * @param upperBound Upper bounds.
+     * @return the value of the objective at the optimum.
+     */
+    private double bobyqb(double[] lowerBound,
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

The cyclomatic complexity of this method is 131. By comparison, 99% of the methods in the CodeGuru reference dataset have a lower cyclomatic complexity. This indicates the method has a high number of decisions and it can make the logic difficult to understand and test. We recommend that you simplify this method or break it into multiple methods.